File: adaptive_repair/data/python_programs/lis.py
import bisect

def lis(arr):
    # 'tails' is an array where tails[k] stores the smallest ending element
    # of an increasing subsequence of length k+1. For example, if tails[0] = 1,
    # it means there's an increasing subsequence of length 1 ending with 1.
    # If tails[1] = 3, it means there's an increasing subsequence of length 2 ending with 3.
    # The 'tails' array is always kept sorted in increasing order.
    tails = []

    for num in arr:
        # Use binary search to find the appropriate position for 'num' in 'tails'.
        # bisect_left returns an insertion point 'idx' such that all elements
        # in tails[:idx] are less than 'num', and all elements in tails[idx:]
        # are greater than or equal to 'num'.
        idx = bisect.bisect_left(tails, num)

        if idx == len(tails):
            # If 'idx' is equal to the current length of 'tails', it means 'num'
            # is greater than all elements currently in 'tails'. This implies 'num'
            # can extend the longest increasing subsequence found so far by one element.
            # So, we append 'num' to 'tails', effectively creating a new longest subsequence.
            tails.append(num)
        else:
            # Otherwise, 'num' is not greater than all elements in 'tails'.
            # 'tails[idx]' is the smallest element in 'tails' that is >= 'num'.
            # By replacing 'tails[idx]' with 'num', we are saying that we found an
            # increasing subsequence of length 'idx + 1' that ends with 'num'.
            # Since 'num' <= 'tails[idx]', this new subsequence ends with a smaller
            # or equal value. A smaller ending value for a subsequence of a given length
            # is always preferable, as it provides more opportunities for future elements
            # to extend it. This operation maintains the sorted property of 'tails'.
            tails[idx] = num

    # The final length of the 'tails' array is the length of the Longest Increasing Subsequence.
    return len(tails)

File: adaptive_repair/data/java_programs/LIS.java
package java_programs;
import java.util.*;

public class LIS {
    public static int lis(int[] arr) {
        if (arr == null || arr.length == 0) {
            return 0;
        }

        // tails[k] stores the smallest tail of an increasing subsequence of length k.
        // The array is 1-indexed for convenience, so its size is arr.length + 1.
        // We only care about tails[1] to tails[longest].
        // The values in tails[1...longest] are always sorted in increasing order.
        int[] tails = new int[arr.length + 1];
        int longest = 0; // Current length of the longest increasing subsequence found

        for (int val : arr) {
            // Perform a binary search to find the correct position for 'val' in the 'tails' array.
            // We are looking for the smallest 'k' such that tails[k] >= val.
            // This 'k' will be the length of the LIS ending with 'val'.
            // Arrays.binarySearch returns:
            //   - index if 'val' is found (meaning tails[index] == val)
            //   - (-(insertion point) - 1) if 'val' is not found.
            //     The 'insertion point' is the index where 'val' would be inserted
            //     to maintain the sorted order. It's also the smallest index 'k'
            //     such that tails[k] >= val.
            // The search range is from index 1 (inclusive) to 'longest + 1' (exclusive).
            int idx = Arrays.binarySearch(tails, 1, longest + 1, val);

            int newLength;
            if (idx < 0) {
                // 'val' was not found. Calculate the insertion point.
                // The insertion point is the length of the LIS that 'val' can form.
                newLength = -(idx + 1);
            } else {
                // 'val' was found at 'idx'. This means tails[idx] == val.
                // We can replace it with 'val' itself (no effective change to the value,
                // but conceptually we update the tail for this length). The length remains 'idx'.
                newLength = idx;
            }

            // Update the 'tails' array at the calculated position.
            tails[newLength] = val;

            // If we found a new longest increasing subsequence, update 'longest'.
            if (newLength > longest) {
                longest = newLength;
            }
        }
        return longest;
    }
}